### PR TITLE
Azure OIDC Support

### PR DESCRIPTION
**Pull request by veochen-octopus**
Adds support for azure OIDC. To be followed by another PR to manage these accounts in the CLI.

[sc-57220]

❗ I had to turn on the feature toggles and use the internal docker image to enable OIDC. Might be best to wait for 2023.4 to go GA first.

### CODE REVIEW

File: pkg/accounts/azure_oidc_account.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	ApplicationID           *uuid.UUID `json:"ClientId" validate:"required"`
+	AuthenticationEndpoint  string     `json:"ActiveDirectoryEndpointBaseUri,omitempty" validate:"required_with=AzureEnvironment,omitempty,uri"`
+	AzureEnvironment        string     `json:"AzureEnvironment,omitempty" validate:"omitempty,oneof=AzureCloud AzureChinaCloud AzureGermanCloud AzureUSGovernment"`
+	ResourceManagerEndpoint string     `json:"ResourceManagementEndpointBaseUri" validate:"required_with=AzureEnvironment,omitempty,uri"`
+	SubscriptionID          *uuid.UUID `json:"SubscriptionNumber" validate:"required"`
+	TenantID                *uuid.UUID `json:"TenantId" validate:"required"`
+	Audience                string     `json:"Audience,omitempty"`
+	DeploymentSubjectKeys   []string   `json:"DeploymentSubjectKeys,omitempty" validate:"omitempty,dive,oneof=space environment project tenant runbook account type'"`
+	HealthCheckSubjectKeys  []string   `json:"HealthCheckSubjectKeys,omitempty" validate:"omitempty,dive,oneof=space account target type'"`
+	AccountTestSubjectKeys  []string   `json:"AccountTestSubjectKeys,omitempty" validate:"omitempty,dive,oneof=space account type'"`
```
Comment (veochen-octopus) [inline pkg/accounts/azure_oidc_account.go:22]: Not sure if validation is necessary on these fields. The API does filter out invalid values but thought these could serve as guidance in the CLI.

File: pkg/accounts/azure/azurewebapp.go [hunk 2]
```diff
@@ -22,7 +23,7 @@ type AzureWebAppSlot struct {
 	ResourceGroup string `json:"ResourceGroup,omitempty"`
 }
 
-func GetWebSites(client client.Client, account *accounts.AzureServicePrincipalAccount) ([]*AzureWebApp, error) {
+func GetWebSites(client client.Client, account accounts.IAccount) ([]*AzureWebApp, error) {
```
Comment (veochen-octopus) [inline pkg/accounts/azure/azurewebapp.go:26]: Easy way to accept both service principal and oidc. Only consumed in the CLI so didn't test.

File: .github/workflows/integration-tests.yml [hunk 3]
```diff
@@ -29,13 +29,15 @@ jobs:
...
-        image: octopusdeploy/octopusdeploy
+        image: docker.packages.octopushq.com/octopusdeploy/octopusdeploy:2024.1.1581
         env:
           ACCEPT_EULA: Y
           DB_CONNECTION_STRING: 'Server=sqlserver;Database=OctopusDeploy;User Id=sa;Password=${{ env.SA_PASSWORD }};'
           ADMIN_API_KEY: ${{ env.ADMIN_API_KEY }}
           ENABLE_USAGE: N
           OCTOPUS_SERVER_BASE64_LICENSE: ${{ secrets.OCTOPUS_SERVER_BASE64_LICENSE }}
+          OCTOPUS__FeatureToggles__OidcAccountsFeatureToggle: true
+          OCTOPUS__FeatureToggles__ServiceAccountOidcIdentitiesFeatureToggle: true
```
Comment (veochen-octopus) [inline .github/workflows/integration-tests.yml:40]: Once 2023.4 is GA we can change all of these back, assuming at that point the feature toggles have also been enabled.

**Review by benPearce1 — APPROVED** [review]
Tested in the go-cli and terrform

**Pull request merged**